repos/keg-cli/src/utils/services/publishService.js

```diff
@@ -0,0 +1,338 @@
+    const message = `Updating ${context} to version ${newVersion}`
+    !dryrun && await runGitCmd(`commit -m \"${message}\"`, repo.location)
+
+    // Push the branch to github
+    publishArgs.step = { number: 9, name: 'git-push' }
+    !dryrun && await runGitCmd(`push ${remote} ${newBranch}`, repo.location)
+  }
+  catch(err){
+    Logger.error(`Error creating git branch`, err.stack)
+    await rollbackChanges(repo, publishArgs)
+  }
+
+  // Add the update to updated, so we know this repo was published
+  return updated.concat([ {...repo, ...publishArgs} ])
+
+}
+
+/**
+ * Runs a set of yarn commands to test, build and publish a repo
+ * @function
+ * @param {Object} repo - Repo object containing meta-data about the current repo
+ * @param {Object} publishContext - Defines how the repo should be published
+ * @param {Object} publishArgs - Define the state or the repo being published
+ * @param {Object} params - Options passed from the command line
+ * 
+ * @returns {Boolean} - True if the repo was published to npm
+ */
+const repoYarnCommands = async (repo, publishContext, publishArgs, params) => {
+
+  const {
+    test,
+    build,
+    publish,
+    access='public',
+  } = publishContext.tasks
+  const { newVersion } = publishArgs
+  const { dryrun } = params 
+
+  try {
+    // Callback when an error is thrown for a repo script
+    const scriptError = script => async () => {
+      Logger.error(`Error running script ${script}`)
+      await rollbackChanges(repo, publishArgs)
+      return false
+    }
+
+    // Run the repos tests
+    publishArgs.step = { number: 2, name: 'test' }
+    logFormal(repo, `${test ? 'Running' : 'Skipping'} yarn test...`)
+    test && await runRepoScript(repo, `test`, scriptError(`test`))
+
+    // Build the repo
+    publishArgs.step = { number: 3, name: 'build' }
+    logFormal(repo, `${build ? 'Running' : 'Skipping'} yarn build...`)
+    build && await runRepoScript(repo, `build`, scriptError(`build`))
+
+    // Publish to NPM
+    publishArgs.step = { number: 4, name: 'publish' }
+    const shouldPublish = publish && !dryrun
+    logFormal(repo, `${shouldPublish ? 'Running' : 'Skipping'} yarn publish...`)
+    const isPublished = shouldPublish && await runRepoScript(repo, `publish --access ${access} --new-version ${newVersion}`, scriptError(`publish`))
+
+    return isPublished
+
+  }
+  catch(err){
+    Logger.error(`Error publishing ${repo.repo}`)
+    await rollbackChanges(repo, publishArgs)
+
+    return false
+  }
+
+}
+
+/**
+ * Copies over new build files to the current repos node_modules
+ * @param {Object} currentRepo 
+ * @param {Array} repos - array of all repos
+ */
+const copyBuildFiles = (currentRepo, repos) => {
+  logFormal(currentRepo, `Copying dependency build files`)
+  repos.map((repo) => {
+    const packageName = get(repo, 'package.name')
+    // check if current repo's deps contain last repo as dependencies
+    if(get(currentRepo, `package.dependencies.${packageName}`)
```

Comment:
yeah i guess this would work too. by the time we get to this function, `yarn install` should have been successful

the one thing I can think of with this approach is that it's not as clear as checking the `package.json` for the dependencies. because there's a possibility that the user can delete that specific node_module for whatever reason, so then it wont get copied over when it should.

I can still update it if you're not convinced, but just wanted to bring it up
it's a very unlikely scenario, but it's a scenario that doesn't exist in the other approach.

what do you think?

Reply:
Yah, I'm good with keeping it as is. This was more a suggestion than anything else. 

Although I don't think we'll run into the case you describe. In your case, `yarn test` and `yarn build` will fail, if it's missing required `node_modules`. 

Either way it's all good.